pkg/controller/petset/pet.go

```diff
@@ -0,0 +1,267 @@
+		glog.Infof("PetSet %v deleting pet %v", pet.parent.Name, pet.pod.Name)
+		return p.petClient.Delete(pet)
+	}
+	glog.Infof("PetSet %v waiting on pet %v to die in %v", pet.parent.Name, realPet.pod.Name, realPet.pod.DeletionTimestamp)
+	return nil
+}
+
+// petClient is a client for managing pets.
+type petClient interface {
+	pvcClient
+	petHealthChecker
+	Delete(*pcb) error
+	Get(*pcb) (*pcb, bool, error)
+	Create(*pcb) error
+	Update(*pcb, *pcb) error
+}
+
+type apiServerPetClient struct {
+	c        *client.Client
+	recorder record.EventRecorder
+	petHealthChecker
+}
+
+func (p *apiServerPetClient) Get(pet *pcb) (*pcb, bool, error) {
+	found := true
+	ns := pet.parent.Namespace
+	pod, err := podClient(p.c, ns).Get(pet.pod.Name)
+	if errors.IsNotFound(err) {
+		found = false
+		err = nil
+	}
+	if err != nil || !found {
+		return nil, found, err
+	}
+	realPet := *pet
+	realPet.pod = pod
+	return &realPet, true, nil
+}
+
+func (p *apiServerPetClient) Delete(pet *pcb) error {
+	err := podClient(p.c, pet.parent.Namespace).Delete(pet.pod.Name, nil)
+	if errors.IsNotFound(err) {
+		err = nil
+	}
+	if err != nil {
+		p.recorder.Eventf(pet.parent, api.EventTypeWarning, "FailedDelete", "pet: %v, error: %v", pet.pod.Name, err)
+	} else {
+		p.recorder.Eventf(pet.parent, api.EventTypeNormal, "SuccessfulDelete", "pet: %v", pet.pod.Name)
+	}
+	return err
+}
+
+func (p *apiServerPetClient) Create(pet *pcb) error {
+	_, err := podClient(p.c, pet.parent.Namespace).Create(pet.pod)
+	if err == nil {
+		p.recorder.Eventf(pet.parent, api.EventTypeNormal, "SuccessfulCreate", "pet: %v", pet.pod.Name)
+	} else {
+		p.recorder.Eventf(pet.parent, api.EventTypeWarning, "FailedCreate", "pet: %v, error: %v", pet.pod.Name, err)
+	}
+	return err
+}
+
+func (p *apiServerPetClient) Update(pet *pcb, expectedPet *pcb) (updateErr error) {
+	var getErr error
+	pc := podClient(p.c, pet.parent.Namespace)
+
+	pod, needsUpdate, err := copyPetID(pet, expectedPet)
+	if err != nil || !needsUpdate {
+		return err
+	}
+	glog.Infof("Resetting pet %v to match PetSet %v spec", pod.Name, pet.parent.Name)
+	for i, p := 0, &pod; ; i++ {
+		_, updateErr = pc.Update(p)
+		if updateErr == nil || i >= updateRetries {
+			return updateErr
+		}
+		if p, getErr = pc.Get(pod.Name); getErr != nil {
+			return getErr
+		}
+	}
+}
+
+func (p *apiServerPetClient) DeletePVCs(pet *pcb) error {
+	// TODO: Implement this when we delete pvcs.
+	return nil
+}
+
+func (p *apiServerPetClient) getPVC(pvcName, pvcNamespace string) (*api.PersistentVolumeClaim, bool, error) {
+	found := true
+	pvc, err := claimClient(p.c, pvcNamespace).Get(pvcName)
+	if errors.IsNotFound(err) {
+		found = false
+	}
+	if err != nil || !found {
+		return nil, found, err
+	}
+	return pvc, true, nil
+}
+
+func (p *apiServerPetClient) createPVC(pvc *api.PersistentVolumeClaim) error {
+	_, err := claimClient(p.c, pvc.Namespace).Create(pvc)
+	return err
+}
+
+func (p *apiServerPetClient) SyncPVCs(pet *pcb) error {
+	errMsg := ""
```

Comment:
Clarify? this is an aggregate error string that maps to errors for pvcs of a single pet. I'm doing a list of errors up the stack anway for pets in a petset, and wanted to avoid a list of lists. Having one error for a single pet sync failure is a little easier.